[SE-0304] Align Task API implementation with the fourth revision of the proposal

DougGregor opened the pull request:
Specific changes are:
* Renaming the task group `async` operations to `addTask`
* Rename `TaskPriority.default` to `TaskPriority.medium`
* Rename `Task.yield()` to `Task.suspend()`
* Add task group `waitForAll()`
* Implement cancellable `Task.sleep(nanoseconds:)`

All changes with the exception of `Task.sleep(nanoseconds:)` are implemented in an ABI-neutral manner.
`Task.sleep(nanoseconds:)` is ABI-additive; it's too much code to want to inline.

Implements rdar://80339563

DougGregor committed:
[SE-0304] Replace the `async` operations in `(Throwing)TaskGroup` with `addTask`

DougGregor committed:
[SE-0304] Rename `TaskPriority.default` to `medium`

DougGregor committed:
[SE-0304] Rename `Task.yield()` to `Task.suspend()`

DougGregor committed:
[SE-0304] Add `(Throwing)?TaskGroup.waitForAll()`

Do this as a staged change to the ABI, introducing an underscored
`@usableFromInline` implementation to the ABI that we can rely on
later, and an `@_alwaysEmitIntoClient` version we can inline now.

DougGregor committed:
[SE-0304] Implement cancellable `Task.sleep(nanoseconds:)`.

benlings commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -623,6 +632,81 @@ extension Task where Success == Never, Failure == Never {
+    // when the task has been cancelled.
+    var cancellationRecord: UnsafeRawPointer? = nil
+    defer {
+      _taskRemoveCancellationHandler(record: cancellationRecord!)
```
Does this remove the cancellation handler too soon? I can see the same thing being done in `withTaskCancellationHandler`, but there it’s deferred until after the operation is awaited. I think the cancellation handler should only be removed in `onWake` or `onCancel`.

benlings replied:
Ah - missed the await on `withUnsafeThrowingContinuation` 🤦 , so it should happen after the sleep has ended.

DougGregor replied:
I switched over to `withTaskCancellationHandler` to be certain that we got the ordering right.

benlings replied:
Thanks - wasn't sure if I'd missed something when I first looked.

benlings commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -623,6 +632,81 @@ extension Task where Success == Never, Failure == Never {
+          // The sleep finished first, which means the the continuation was
+          // called by the operation. All we have left to do is deallocate the
+          // flag word storage.
+          wordPtr.deallocate()
```
Does this leak `wordPtr` if the task isn’t cancelled?

DougGregor replied:
It does! I've fixed this in my slightly-less-broken-but-still-somehow-wrong version.

benlings commented on test/Concurrency/Runtime/async_task_sleep_cancel.swift:
```diff
@@ -0,0 +1,70 @@
+    let sleepyTask = Task {
+      try await Task.sleep(nanoseconds: UInt64(pause))
+    }
+    sleepyTask.cancel()
```
I think this also needs a test for if the task is cancelled after it has started.

DougGregor replied:
Right, we need both.

benlings reviewed (commented):
Hope an unsolicited review is OK

DougGregor commented:
@swift-ci please test

phausler commented on test/Concurrency/Runtime/async_task_sleep_cancel.swift:
```diff
@@ -0,0 +1,110 @@
+
+@available(SwiftStdlib 5.5, *)
+@main struct Main {
+  static let pause = 500_000_000 // 500ms
```
it might also be good to test the 0 case of a pause

DougGregor replied:
Great point, I'm adding it in a subsequent commit.

phausler commented on stdlib/public/Concurrency/TaskSleep.swift:
```diff
@@ -0,0 +1,144 @@
+        }
+      } onCancel: {
+        let continuationWord = continuationPtr.pointee
+        if UInt(continuationWord) != 0 {
```
is this safe from here on out if another thread cancels just after the `checkCancellation` at line 84?

DougGregor replied:
So, there's an existing race in `withTaskCancellationHandler` that also gets replicated here. When we address that (@ktoso was looking into it), this will also be fixed.

ktoso replied:
I'll get to it this week. I think we'll end up having to "check again" after having stored a cancellation record basically.

phausler commented on stdlib/public/Concurrency/TaskSleep.swift:
```diff
@@ -0,0 +1,144 @@
+    try checkCancellation()
+
+    // Allocate storage for the flag word and continuation.
+    let wordPtr = UnsafeMutablePointer<Builtin.Word>.allocate(capacity: 2)
```
this word does not escape the context of this function right? it might be interesting to actually use a stack allocation instead of heap allocating it.

else wise if it does escape then this probably should be ref counted right?

DougGregor replied:
It does escape the context of this function: when the task is cancelled, the timed "wake" task has yet to be scheduled, and needs to check whether it finished first or not. In the case where it gets called but didn't finish first, it handles the deallocation.

No need for reference counting; the deallocation points are well-defined.

DougGregor committed:
[SE-0304] Clarify documentation of Task.sleep(nanoseconds:) slightly

DougGregor committed:
[SE-0304] Add `UnsafeCurrentTask.cancel()`.

DougGregor committed:
Add a test for `Task.sleep(nanoseconds: 0)`.

DougGregor committed:
Remove `UnownedJob.run()`. It never did anything anyway

DougGregor committed:
Reimplement `Task.sleep(nanoseconds:)` without the raciness.

The prior implementation of `Task.sleep()` effectively had two
different atomic words to capture the state, which could lead to cases
where cancelling before a sleep operation started would fail to
throw `CancellationError`. Reimplement the logic for the cancellable
sleep with a more traditional lock-free approach by
packing all of the state information into a single word, where we
always load, figure out what to do, then compare-and-swap.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test Windows
